### ISSUE TITLE
Improving Accuracy of 3D Hand Keypoint Coordinates Relative to Camera Using Depth Map Correction

### ISSUE THREAD

**Issue by Vincia-Jun** [geopavlakos/hamer#118]
Dear contributors,

Thank you very much for your valuable work. I'm working on obtaining accurate 3D coordinates of 21 hand keypoints relative to the camera. Based on the discussion in #30, I understand that the coordinates can be calculated as:

**pred_keypoints_3d** + **pred_cam_t_full**

Here's my implementation:
```python
hand_points = np.array([[ 0.0957,  0.0064,  0.0062],
                        [ 0.0617, -0.0088,  0.0161],
                        [ 0.0316, -0.0027,  0.0181],
                        [ 0.0058,  0.0055,  0.0193],
                        [-0.0242,  0.0166,  0.0346],
                        [ 0.0589, -0.0536,  0.0631],
                        [ 0.0477, -0.0650,  0.0920],
                        [ 0.0343, -0.0600,  0.1090],
                        [ 0.0185, -0.0496,  0.1257],
                        [ 0.0776, -0.0505,  0.0797],
                        [ 0.0600, -0.0487,  0.1059],
                        [ 0.0453, -0.0404,  0.1220],
                        [ 0.0259, -0.0304,  0.1357],
                        [ 0.0958, -0.0314,  0.0836],
                        [ 0.0762, -0.0354,  0.1044],
                        [ 0.0620, -0.0272,  0.1230],
                        [ 0.0423, -0.0182,  0.1358],
                        [ 0.1068, -0.0127,  0.0851],
                        [ 0.0904, -0.0178,  0.0974],
                        [ 0.0776, -0.0164,  0.1113],
                        [ 0.0606, -0.0118,  0.1207]])

pred_cam_t_full = np.array([0.24308263, 0.06009501, 0.262 ])
hand_points += pred_cam_t_full
```

Since **pred_cam_t_full** depends on the **scaled_focal_length**, I've modified it to use my camera's exact intrinsic parameters:
```python
scaled_focal_length = 261
pred_cam_t_full = cam_crop_to_full(pred_cam, box_center, box_size, img_size, scaled_focal_length).detach().cpu().numpy()
```

While the resulting coordinates (after adding hand_points and pred_cam_t_full) represent the estimated positions relative to the original camera, I'm observing position errors ranging from 5-20cm. I suspect this is due to limitations in monocular RGB-based estimation.

To address this, I'm attempting to use a depth map to correct these position errors. I have obtained the average depth of the hand relative to the camera (real_hand_depth) and am considering two correction approaches:

**Way 1:**
```python
pred_cam_t_full[2] = real_hand_depth
```

**Way 2:**
```python
lambda = real_hand_depth / pred_cam_t_full[2]
pred_cam_t_full *= lambda
```

Could you advise which approach is more reasonable? Alternatively, are there better methods to utilize depth maps for achieving more accurate 3D hand keypoint coordinates relative to the camera?

Any insights you could provide would be greatly appreciated.

Best regards,
Jim

**Comment by shinshiner** [comment]
Same problem, refer to the newest issue. I'm trying to use ICP to finetune the estimated pose and not working good enough. Is there any work using RGBD and can obtain accurate translation estimation?
